distribution/lib/Standard/Base/0.0.0-dev/src/Random.enso

```diff
@@ -1,99 +1,343 @@
+         Generate a uuid.
+
+             import Standard.Base.Random.Random
+
+             uuid = Random.uuid
+    uuid : Text
+    uuid = Random_Generator.global_random_generator.uuid
 
     ## GROUP Random
-       Gets the next random Float from a normal distribution with mean 0 and std-dev 1.
-    gaussian : Float
-    gaussian self =
-        self.java_random.nextGaussian
+       Returns a new `Vector` containing a random sample of the input vector,
+       with or without replacement.
+
+       Arguments:
+       - vector: The `Vector` to take items from.
+       - count: The number of items to return.
+       - with_replacement: If false, each value in the `Vector` can only be
+         selected once during a single call; the returned values will contain no
+         duplicates. If true, values from the `Vector` can be returned multiple
+         times and duplicates are possible.
+
+       ? Selection without replacement
+
+         If `with_replacement` is false, and `count` is equal to the size of the
+         collection, a random permutation of the collection is returned.
+
+         If `with_replacement` is false, and `count` is greater than the size of
+         the collection, an `Illegal_Argument` error is thrown. (If `count` is
+         greater than the size of the collection, `items` will run out of values
+         to choose from, and will not be able to return enough values.)
+
+       > Example
+         Sample a vector.
+
+             import Standard.Base.Random.Random
+
+             items = Random.items [1, 2, 3, 4]
+    items : (Vector Any | Array Any) -> Integer -> Boolean -> Vector Any ! Illegal_Argument
+    items (v : Vector Any | Array Any) (count:Integer=1) (with_replacement:Boolean=False) = Random_Generator.global_random_generator.items v count with_replacement
 
     ## GROUP Random
-       Gets the next random integer between min (inclusive) and max (exclusive).
+       Returns a random permutation of the input `Vector`.
 
        Arguments:
-        - min: the minimum value (inclusive) of the random integer.
-        - max: the maximum value (exclusive) of the random integer.
-    integer : Integer -> Integer -> Integer
-    integer self min=0 max=100 =
+       - vector: The `Vector` to permute.
+
+       > Example
+         Permute a vector.
+
+             import Standard.Base.Random.Random
+
+             permuted = Random.permute [1, 2, 3, 4]
+    permute : Vector Any -> Vector Any
+    permute (v:Vector) = Random_Generator.global_random_generator.permute v
+
+# PRIVATE
+type Random_Generator
+    ## PRIVATE
+       Create a new rng with the given seed.
+    new : Integer -> Random_Generator
+    new (seed:Integer) = Random_Generator.Value (RandomInstanceHolder.new seed)
+
+    ## PRIVATE
+    Value (random_instance_holder:RandomInstanceHolder)
+
+    ## PRIVATE
+    global_random_generator : Random_Generator
+    global_random_generator = Random_Generator.Value RandomInstanceHolder.SINGLETON
```

Comment:
Well, it still creates new `Random_Generator.Value` with each call.

Reply:
Is there a way to avoid that?

Reply:
> Well, it still creates new `Random_Generator.Value` with each call.

But now they are referentially transparent - both instances behave exactly the same. So I guess there's no harm in that.

> Is there a way to avoid that?

Currently, not really. We would have to cache the Enso instance in a Java field to make sure it is the same. But that sounds like total overkill.

Ideally, I think it would be good to have an ability to define _constants_ in Enso. Global constants that are computed only once.

Yes, they can be used to implement global state which is often rather yucky. But they can also be a blessing - e.g. we currently have a lot of global immutable singleton instances like of `Map`s (e.g. see `Postgres_Type_Mapping` - e.g. `complex_types_map` or `simple_types_map`, `SQLite_Type_Mapping`). Currently, these maps are constructed each time they are accessed. This is quite a lot of unnecessary work happening. It happens rarely enough that it does not make performance awful, but it just feels bad to recompute this all the time if we could easily store it.

@JaroslavTulach @jdunkerley do you think we should try to add such a feature? 
I know it's probably not a high priority thing because it does not give us much noticeable gains - if we really need such a static field, we can already do it by caching the value in a Java static field - that is just rather ugly solution, but it works. 
At some point though, it would be nice for the Enso language to provide such a capability 'natively'.

Reply:
> Is there a way to avoid that?

Yes, with following trick:
```java
static singleton(java.util.function.Function<Object,Object> factory) {
  if (SINGLETON == null) {
    SINGLETON = factory.apply(new RandomInstanceHolder());
  }
  return SINGLETON;
}
```
and then following Enso code:
```ruby
RandomInstanceHolder.singleton Random_Generator.Value
```

> define constants in Enso

Technically similar to lazy atom field, but on global scope.

Reply:
> > define constants in Enso
> 
> Technically similar to lazy atom field, but on global scope.

Yeah, exactly.

Reply:
Done